backend/src/github-util.ts

```diff
@@ -0,0 +1,283 @@
+  const octokit = new Octokit();
+
+  // non-review comments are classified as "issue" comments
+  return octokit.rest.issues
+    .listComments({ ...pull, issue_number: pull.pull_number })
+    .then((res) => {
+      const comments = res.data;
+      return comments.map(
+        (comment): Comment => ({
+          commentUrl: comment.html_url,
+          createdBy: comment.user?.login || '',
+          createdAt: Date.parse(comment.created_at),
+          content: comment.body || ''
+        })
+      );
+    });
+};
+
+/** Returns `comments` created by `username` between `start_time` and `end_time`.
+ *  Raises an error if no comment satisfies these conditions. */
+const filterComments = (
+  comments: Comment[],
+  username: string,
+  startTime: number,
+  endTime: number
+): Comment[] => {
+  if (!comments || !comments.length) {
+    throw new Error(`No review comments in this PR.`);
+  }
+
+  let eligibleComments: Comment[] = [];
+
+  // comments made by user
+  eligibleComments = comments.filter((comment) => comment.createdBy === username);
+  if (!eligibleComments || !eligibleComments.length) {
+    throw new Error(`No review comments made by user ${username} in this PR.`);
+  }
+
+  // comments made in date range
+  eligibleComments = eligibleComments.filter(
+    (comment) => startTime <= comment.createdAt && comment.createdAt <= endTime
+  );
+  if (!eligibleComments || !eligibleComments.length) {
+    const startDate = new Date(startTime).toDateString();
+    const endDate = new Date(endTime).toDateString();
+    throw new Error(
+      `No review comments made by user ${username} between ${startDate} and ${endDate}.`
+    );
+  }
+
+  return eligibleComments;
+};
+
+/** Retrieves information about `pull_request` and its review comments. */
+const getReviewedPR = async (pull: PullRequest): Promise<ReviewedPR> => {
+  const octokit = new Octokit();
+
+  // get information about a PR and its review comments
+  // cannot get both with a single api call
+  return Promise.all([octokit.rest.pulls.get(pull), getReviewComments(pull)]).then(
+    ([pr, comments]) => ({
+      url: pr.data.html_url,
+      createdBy: pr.data.user?.login || '',
+      comments
+    })
+  );
+};
+
+/** Retrieves information about opened PR `pull_request`. */
+const getOpenedPR = async (pull: PullRequest): Promise<OpenedPR> => {
+  const octokit = new Octokit();
+
+  return Promise.all([octokit.rest.pulls.get(pull), getNonReviewComments(pull)]).then(
+    ([pr, comments]) => {
+      // get any comments by dti bot
+      const botComments = filterComments(comments, 'dti-github-bot', 0, Number.MAX_SAFE_INTEGER);
+
+      let diffSize = 0;
+      botComments.forEach((comment) => {
+        const match = comment.content.match(/.*\[diff-counting\] Significant lines: ([0-9]+).*/);
+        if (match == null) {
+          throw new Error(`No significant line count by dti-github-bot for PR ${pr.data.url}.`);
+        }
```

Comment:
hmm i think we need to be careful about if the big diff bot hasn't finished calculating the diff yet (i.e. the comment isn't there yet)

Reply:
might be fine to leave as it is for now since using diff size to determine if PR is valid is only temporary